feat: argo rollouts charm

IronCore864 opened the pull request.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,199 @@
+    """Charmed Operator for Argo Rollouts."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
```
The tutorial uses this pattern and a lot of charms do too (probably partly as a result) but @benhoyt and I generally prefer (and encourage) `(self, framework)` and then `super().__init__(framework)` and leaving off the `self.` below.

It'd be interesting to get your opinion on this too!

tonyandrewmeyer commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+ops >= 2.10
+lightkube
+glob2
+jinja2
```
Are you using jinja (directly) anywhere?

IronCore864 replied:
When loading yaml with lightkube, if a context is given, the stream is considered a `jinja2` template and the context will be used during templating. This is an optional feature, and lightkube doesn't import jinja2 in its requirements, so it must stay here.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,209 @@
+from lightkube.core.exceptions import ApiError
+import ops
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.testing import Harness
```
We're trying to encourage `import ops.testing` and then `harness = ops.testing.Harness(...)`.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,209 @@
+    @patch.object(ArgoRolloutsOperatorCharm, "_create_kubernetes_resources")
+    def test_install_event_successful(self, create):
+        self.harness.charm.on.install.emit()
+        self.assertIn(
+            ("status_set", "maintenance", "creating kubernetes resources", {"is_app": False}),
+            self.harness._get_backend_calls(),
+        )
```
I think here it would be simpler to just assert on the status (like you do in the test above) rather than on testing implementation details.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,209 @@
+            self.harness.charm.on.install.emit()
+            self.assertTrue(len(logs) > 0)
+
+        self.assertIn(
+            ("status_set", "maintenance", "creating kubernetes resources", {"is_app": False}),
+            self.harness._get_backend_calls(),
+        )
```
Harness doesn't really have a way to look at the status history (the other testing framework, Scenario, does, for what it's worth). I don't think tests really ought to be accessing private methods of the Harness, though.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,209 @@
+            except ApiError:
+                self.assertIn("failed to delete resource:", ";".join(logs.output))
+
+    def test_evaluate_argo_rollouts_status_no_service(self):
+        self.harness.charm.unit.status = WaitingStatus()
+        self.harness.charm._evaluate_argo_rollouts_status()
+        assert self.harness.charm.unit.status == WaitingStatus()
```
A problem here is that `_evaluate_argo_rollouts_status` could do nothing at all and this test would still pass. It's probably better to set the status to something else first rather than waiting.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,209 @@
+        harness.begin()
+        self.assertEqual(harness.charm._namespace, "test")
+        mock.assert_called_with("/var/run/secrets/kubernetes.io/serviceaccount/namespace", "r")
+        harness.cleanup()
```
This should be either in a `tearDown` or done with `addCleanup` (or in a finally block), otherwise it won't happen if either of the asserts fail.

tonyandrewmeyer reviewed (commented):
This is a really great first charm! I don't know Argo Rollouts so can't tell if the specifics of that are working properly, but I'm assuming they are. This turned out pretty straightforward (no need for actions, config, secrets, etc) which is a good choice for a first charm (better than my choice was!).

I've added a few comments that we can talk over on Monday with Ben. A bunch of these are related to style choices that are pretty recent (some pushed by me!) and not reflected yet in the tutorial or across a lot of charms.

benhoyt commented on src/charm.py:
```diff
@@ -0,0 +1,199 @@
+        if services != new_layer["services"]:
+            self.container.add_layer("argo-rollouts", self._pebble_layer, combine=True)
+            logger.info("Added updated layer 'argo_rollouts' to Pebble plan")
+            self.container.restart(self.pebble_service_name)
```
Yeah, `replan` is definitely the preferred way, as Pebble will try to only restart the service if the plan has changed in meaningful way.

benhoyt reviewed (commented):
Just a couple comments in addition to Tony's. We'll chat further on Monday!

benhoyt commented on src/charm.py:
```diff
@@ -0,0 +1,199 @@
+
+    def _handle_ports(self):
+        metrics_port = 8090
+        self.unit.set_ports(metrics_port)
```
For example, `METRICS_PORT = 8090` at top level.

benhoyt commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+ops >= 2.10
+lightkube
+glob2
```
Maybe try https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob or https://docs.python.org/3/library/glob.html

benhoyt commented on tests/integration/test_charm.py:
```diff
@@ -0,0 +1,96 @@
+    logger.info("Building charm...")
+    charm = await ops_test.build_charm(".")
+
+    bundles = [Path("tests/data/charm.yaml")]
```
Tiexin to add code comment: it's because bundles support architecture?

IronCore864 replied:
I have figured out how to pass constraints to ops_test.model.deploy, I will remove the bundle part.

IronCore864 commented:
I resolved all comments, fixed them (in separated commits for classification), and added some manual test doc to validate that argo rollouts is working properly.

Tests all passed on both arm64 and amd64.

benhoyt commented on src/charm.py:
```diff
@@ -7,48 +7,46 @@
-        self.framework.observe(self.on.install, self._on_install_or_upgrade)
-        self.framework.observe(self.on.upgrade_charm, self._on_install_or_upgrade)
-        self.framework.observe(self.on.remove, self._on_remove)
+        framework.observe(self.on.argo_rollouts_pebble_ready, self._install_and_restarty)
```
Is the "y" on the end a typo?
```suggestion
        framework.observe(self.on.argo_rollouts_pebble_ready, self._install_and_restart)
```

benhoyt commented on charmcraft.yaml:
```diff
@@ -0,0 +1,64 @@
+# environment constraint
+# assuming the latest major Juju version and a Kubernetes-type cloud
+# at the moment, the latest juju version is 3.4
+# juju 3.1 is used here because 3.4 can't bootstrap a contoller successfully on linux/arm64
```
```suggestion
# juju 3.1 is used here because 3.4 can't bootstrap a controller successfully on linux/arm64
```

benhoyt commented on tox.ini:
```diff
@@ -0,0 +1,88 @@
+[testenv:static]
+description = Run static type checks
+deps =
+    pyright
```
I'm hitting a Pyright bug where it's failing on 1.1.352 and later (current version is 1.1.353). Pinning it to 1.1.351 fixed it for me.

https://github.com/RobertCraigie/pyright-python/issues/254

benhoyt reviewed (changes requested):
I think this is in a good state for a v1, and we can always improve further from there as needed! However, just a couple of tox/testing things:

Various `tox` commands fail right now:

- `tox -e format` reformats imports
- `tox -e lint` has codespell errors ("contoller")
- `tox -e static` has an issue due to a Pyright bug (see below) -- let's pin the version
- `tox -e unit` works but some of the tests are slow, when IMO they should be instant -- is that because of the time.sleep in the failure case? If so I think we should mock time.sleep with a no-op as good practice.

tonyandrewmeyer commented:
> I think this is in a good state for a v1, and we can always improve further from there as needed! However, just a couple of tox/testing things:
> 
> Various `tox` commands fail right now:
> 
> * `tox -e format` reformats imports
> * `tox -e lint` has codespell errors ("contoller")
> * `tox -e static` has an issue due to a Pyright bug (see below) -- let's pin the version
> * `tox -e unit` works but some of the tests are slow, when IMO they should be instant -- is that because of the time.sleep in the failure case? If so I think we should mock time.sleep with a no-op as good practice.

For what it's worth, I really like using the [pre-commit framework](https://pre-commit.com/) with Git pre-commit hooks to catch most of these (e.g. [ruff](https://github.com/astral-sh/ruff-pre-commit)). I'm on the fence about unit tests in the check, but the rest are definitely worthwhile.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,201 @@
+            return (
+                ops.ActiveStatus()
+                if service and service.is_running()
+                else ops.WaitingStatus("Waiting for Argo Rollouts service")
```
@benhoyt here's a good example of `WaitingStatus` use vs. our docs, as came up in another recent PR. To me, the three uses of `WaitingStatus` here are all perfectly reasonable, and make more sense than `BlockedStatus`, which seems like the only alternative. However, the ops [docs](https://ops.readthedocs.io/en/latest/#ops.WaitingStatus) say to use this when specifically waiting for related applications. Thoughts?

benhoyt replied:
Discussed with Juju team. See [Matrix thread](https://matrix.to/#/!xzmWHtGpPfVCXKivIh:ubuntu.com/$g7l28jPqhxjJOlz5EOQuPWw_2H4PbuZ4OqNkUzxDbhg) and [docstring tweak](https://github.com/canonical/operator/pull/1148).

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,201 @@
+    def version(self) -> str:
+        """Argo Rollouts controller's version."""
+        try:
+            self.container.get_services(self.pebble_service_name)
```
I assume you're doing this to give the specific "pebble not ready" error, but it seems a bit odd to call and not use the result. If you don't have this then you'll still catch the exception, which would be some sort of failure to connect to the endpoint, I assume. It seems like it would be cleaner to let that happen.

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,201 @@
+        return ops.pebble.Layer(pebble_layer)
+
+    def _on_remove(self, event):
+        self.unit.status = ops.MaintenanceStatus("deleting kubernetes resources")
```
You don't change back to `ActiveStatus` after this, although it's a bit of a moot point since the unit is about to vanish anyway. So it's probably fine, just reads a little odd.

tonyandrewmeyer replied:
Although TIL that apparently [you are *meant* to move to MaintenanceStatus](https://juju.is/docs/sdk/constructs#heading--statuses) on stop, so this is actually fine (although a wee bit later).

tonyandrewmeyer commented on src/charm.py:
```diff
@@ -0,0 +1,201 @@
+                    except ApiError:
+                        logger.debug("failed to delete resource: %s.", resource.to_dict())
+                        raise
+        return True
```
This probably isn't needed - you either return `True` or raise an exception, so there's really only one possible return value.

tonyandrewmeyer commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,192 @@
+
+
+class TestCharmNamespaceProperty(unittest.TestCase):
+    def tearDown(self):
+        harness = ops.testing.Harness(ArgoRolloutsCharm)
+        harness.cleanup()
```
This is cleaning up the harness just created, rather than the one that's created in the test case below.

The typical pattern with unittest is to create the harness in `setUp` (like in the class above) and call `addCleanup` there. If you want to use `tearDown`, you'd still create the harness object in `setUp` and then have:

```python
def tearDown(self):
    self.harness.cleanup()
```

Or with pytest, the common pattern is a fixture, like:

```python
@pytest.fixture()
def harness():
    harness = ops.testing.Harness(MyCharm)
    yield harness
    harness.cleanup()
```

tonyandrewmeyer reviewed (approved):
A few small comments, but looking great!

IronCore864 commented:
> > I think this is in a good state for a v1, and we can always improve further from there as needed! However, just a couple of tox/testing things:
> > Various `tox` commands fail right now:
> > 
> > * `tox -e format` reformats imports
> > * `tox -e lint` has codespell errors ("contoller")
> > * `tox -e static` has an issue due to a Pyright bug (see below) -- let's pin the version
> > * `tox -e unit` works but some of the tests are slow, when IMO they should be instant -- is that because of the time.sleep in the failure case? If so I think we should mock time.sleep with a no-op as good practice.
> 
> For what it's worth, I really like using the [pre-commit framework](https://pre-commit.com/) with Git pre-commit hooks to catch most of these (e.g. [ruff](https://github.com/astral-sh/ruff-pre-commit)). I'm on the fence about unit tests in the check, but the rest are definitely worthwhile.

I have tried my best to fix lint and static errors; some are related to the third-party lightkube lib.

tonyandrewmeyer commented:
> I have tried my best to fix lint and static errors; some are related to the third-party lightkube lib.

tox shouldn't be linting/static checking outside of the `src` and `tests` directories. If it's in those places but the library uses a bad name or something like that, then that's a perfectly valid case for using a [noqa directive](https://docs.astral.sh/ruff/linter/#error-suppression).

IronCore864 commented:
> I think this is in a good state for a v1, and we can always improve further from there as needed! However, just a couple of tox/testing things:
> 
> Various `tox` commands fail right now:
> 
> * `tox -e format` reformats imports
> * `tox -e lint` has codespell errors ("contoller")
> * `tox -e static` has an issue due to a Pyright bug (see below) -- let's pin the version
> * `tox -e unit` works but some of the tests are slow, when IMO they should be instant -- is that because of the time.sleep in the failure case? If so I think we should mock time.sleep with a no-op as good practice.

After some investigation, I don't think the UT has anything to do with the time.sleep, which is only used for retry.

I figured that it's because the create k8s resource test needs to load the templates, which are big files, so it's a bit slow.

benhoyt commented:
Thanks, looks good. I think it's worth making `tox` after a checkout pass cleanly. It looks like lightkube is typed (has a py.typed, [added by our very own Jon Seager](https://github.com/gtsystem/lightkube/pull/13)), but has some incorrect typing in `Client.__init__` for example. It works slightly better with MyPy, but even there there are issues.

So to work around the poor typing in lightkube, I suggest we just sprinkle a few `# type: ignore`s around the place to get it to succeed. You also need one on the `ops.main()` call. This passes for me:

```
diff --git a/src/charm.py b/src/charm.py
index f19af41..b32d492 100755
--- a/src/charm.py
+++ b/src/charm.py
@@ -18,7 +18,7 @@ import requests
 from charms.grafana_k8s.v0.grafana_dashboard import GrafanaDashboardProvider
 from charms.loki_k8s.v0.loki_push_api import LogProxyConsumer
 from charms.prometheus_k8s.v0.prometheus_scrape import MetricsEndpointProvider
-from lightkube import Client, codecs
+from lightkube import Client, codecs  # type: ignore
 from lightkube.core.exceptions import ApiError
 
 logger = logging.getLogger(__name__)
@@ -63,7 +63,7 @@ class ArgoRolloutsCharm(ops.CharmBase):
             return f.read().strip()
 
     def _create_kubernetes_resources(self) -> bool:
-        client = Client(field_manager="argo-rollouts-operator-manager")
+        client = Client(field_manager="argo-rollouts-operator-manager")  # type: ignore
         for manifest in glob("src/templates/*.yaml.j2"):
             with open(manifest) as f:
                 for resource in codecs.load_all_yaml(f, context=self._context):
@@ -151,18 +151,18 @@ class ArgoRolloutsCharm(ops.CharmBase):
             logger.exception("kubernetes API error, resource deletion failed")
 
     def _delete_kubernetes_resources(self) -> bool:
-        client = Client()
+        client = Client()  # type: ignore
         for manifest in glob("src/templates/*.yaml.j2"):
             with open(manifest) as f:
                 for resource in codecs.load_all_yaml(f, context=self._context):
                     try:
                         if resource.metadata and not resource.metadata.namespace:
-                            client.delete(resource.__class__, resource.metadata.name)
+                            client.delete(resource.__class__, resource.metadata.name)  # type: ignore
                         else:
                             client.delete(
-                                resource.__class__,
-                                name=resource.metadata.name,
-                                namespace=resource.metadata.namespace,
+                                resource.__class__,  # type: ignore
+                                name=resource.metadata.name,  # type: ignore
+                                namespace=resource.metadata.namespace,  # type: ignore
                             )
                     except ApiError:
                         logger.debug("failed to delete resource: %s.", resource.to_dict())
@@ -200,4 +200,4 @@ class ArgoRolloutsCharm(ops.CharmBase):
 
 
 if __name__ == "__main__":
-    ops.main(ArgoRolloutsCharm)
+    ops.main(ArgoRolloutsCharm)  # type: ignore
```

IronCore864 commented:
> Thanks, looks good. I think it's worth making `tox` after a checkout pass cleanly. It looks like lightkube is typed (has a py.typed, [added by our very own Jon Seager](https://github.com/gtsystem/lightkube/pull/13)), but has some incorrect typing in `Client.__init__` for example. It works slightly better with MyPy, but even there there are issues.
> 
> So to work around the poor typing in lightkube, I suggest we just sprinkle a few `# type: ignore`s around the place to get it to succeed. You also need one on the `ops.main()` call. This passes for me:
> 
> ```diff
> diff --git a/src/charm.py b/src/charm.py
> index f19af41..b32d492 100755
> --- a/src/charm.py
> +++ b/src/charm.py
> @@ -18,7 +18,7 @@ import requests
>  from charms.grafana_k8s.v0.grafana_dashboard import GrafanaDashboardProvider
>  from charms.loki_k8s.v0.loki_push_api import LogProxyConsumer
>  from charms.prometheus_k8s.v0.prometheus_scrape import MetricsEndpointProvider
> -from lightkube import Client, codecs
> +from lightkube import Client, codecs  # type: ignore
>  from lightkube.core.exceptions import ApiError
>  
>  logger = logging.getLogger(__name__)
> @@ -63,7 +63,7 @@ class ArgoRolloutsCharm(ops.CharmBase):
>              return f.read().strip()
>  
>      def _create_kubernetes_resources(self) -> bool:
> -        client = Client(field_manager="argo-rollouts-operator-manager")
> +        client = Client(field_manager="argo-rollouts-operator-manager")  # type: ignore
>          for manifest in glob("src/templates/*.yaml.j2"):
>              with open(manifest) as f:
>                  for resource in codecs.load_all_yaml(f, context=self._context):
> @@ -151,18 +151,18 @@ class ArgoRolloutsCharm(ops.CharmBase):
>              logger.exception("kubernetes API error, resource deletion failed")
>  
>      def _delete_kubernetes_resources(self) -> bool:
> -        client = Client()
> +        client = Client()  # type: ignore
>          for manifest in glob("src/templates/*.yaml.j2"):
>              with open(manifest) as f:
>                  for resource in codecs.load_all_yaml(f, context=self._context):
>                      try:
>                          if resource.metadata and not resource.metadata.namespace:
> -                            client.delete(resource.__class__, resource.metadata.name)
> +                            client.delete(resource.__class__, resource.metadata.name)  # type: ignore
>                          else:
>                              client.delete(
> -                                resource.__class__,
> -                                name=resource.metadata.name,
> -                                namespace=resource.metadata.namespace,
> +                                resource.__class__,  # type: ignore
> +                                name=resource.metadata.name,  # type: ignore
> +                                namespace=resource.metadata.namespace,  # type: ignore
>                              )
>                      except ApiError:
>                          logger.debug("failed to delete resource: %s.", resource.to_dict())
> @@ -200,4 +200,4 @@ class ArgoRolloutsCharm(ops.CharmBase):
>  
>  
>  if __name__ == "__main__":
> -    ops.main(ArgoRolloutsCharm)
> +    ops.main(ArgoRolloutsCharm)  # type: ignore
> ```

Updated. I can't work out how to let `ops.main(ArgoRolloutsCharm)` pass without ignoring it, so I used `# type: ignore` anyway at last...
